hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/io/compress/zlib/BuiltInGzipCompressor.java

```diff
@@ -0,0 +1,270 @@
+    int numAvailBytes = 0;
+
+    // If we are not within uncompressed data yet, output the header.
+    if (state != BuiltInGzipDecompressor.GzipStateLabel.INFLATE_STREAM) {
+      if (userBufLen <= 0) {
+        return numAvailBytes;
+      }
+
+      int outputHeaderSize = writeHeader(b, off, len);
+      headerBytesWritten += outputHeaderSize;
+
+      // Completes header output.
+      if (headerOff == GZIP_HEADER_LEN) {
+        state = BuiltInGzipDecompressor.GzipStateLabel.INFLATE_STREAM;
+      }
+
+      numAvailBytes += outputHeaderSize;
+
+      if (outputHeaderSize == len) {
+        return numAvailBytes;
+      }
+
+      off += outputHeaderSize;
+      len -= outputHeaderSize;
+    }
+
+    if (state == BuiltInGzipDecompressor.GzipStateLabel.INFLATE_STREAM) {
+      // hand off user data (or what's left of it) to Deflater--but note that
+      // Deflater may not have consumed all of previous bufferload, in which case
+      // userBufLen will be zero
+      if (userBufLen > 0) {
+        deflater.setInput(userBuf, userBufOff, userBufLen);
+
+        crc.update(userBuf, userBufOff, userBufLen);  // CRC-32 is on uncompressed data
+
+        currentBufLen = userBufLen;
+        userBufOff += userBufLen;
+        userBufLen = 0;
+      }
+
+
+      // now compress it into b[]
+      int deflated = deflater.deflate(b, off, len);
+
+      numAvailBytes += deflated;
+      off += deflated;
+      len -= deflated;
+
+      // All current input are processed. Going to output trailer.
+      if (deflater.finished()) {
+        state = BuiltInGzipDecompressor.GzipStateLabel.TRAILER_CRC;
+        fillTrailer();
+      } else {
+        return numAvailBytes;
+      }
+    }
+
+    numAvailBytes += writeTrailer(b, off, len);
+    trailerBytesWritten += numAvailBytes;
+
+    if (trailerOff == GZIP_TRAILER_LEN) {
+      state = BuiltInGzipDecompressor.GzipStateLabel.FINISHED;
+      currentBufLen = 0;
+      headerOff = trailerOff = 0;
+    }
+
+    return numAvailBytes;
+  }
+
+  @Override
+  public long getBytesRead() {
+    return deflater.getTotalIn();
+  }
+
+  @Override
+  public long getBytesWritten() {
+    return headerBytesWritten + deflater.getTotalOut() + trailerBytesWritten;
+  }
+
+  @Override
+  public void end() {
+    deflater.end();
+  }
+
+  @Override
+  public void finish() {
+    deflater.finish();
+    state = BuiltInGzipDecompressor.GzipStateLabel.FINISHED;
+  }
+
+  private void init(Configuration conf) {
+    ZlibCompressor.CompressionLevel level = ZlibFactory.getCompressionLevel(conf);
+    ZlibCompressor.CompressionStrategy strategy = ZlibFactory.getCompressionStrategy(conf);
+
+    // 'true' (nowrap) => Deflater will handle raw deflate stream only
+    deflater = new Deflater(level.compressionLevel(), true);
+    deflater.setStrategy(strategy.compressionStrategy());
+
+    state = BuiltInGzipDecompressor.GzipStateLabel.HEADER_BASIC;
+
+    GZIP_TRAILER = new byte[]{0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00};
```

Comment:
not sure if this is necessary, since the array will be overwritten.

Reply:
> I tried to run the tests locally but found one failure:
> 
> ```
> [INFO] Running org.apache.hadoop.io.compress.TestCodec
> [ERROR] Tests run: 28, Failures: 1, Errors: 0, Skipped: 4, Time elapsed: 30.735 s <<< FAILURE! - in org.apache.hadoop.io.compress.TestCodec
> [ERROR] testCodecPoolCompressorReinit(org.apache.hadoop.io.compress.TestCodec)  Time elapsed: 0.005 s  <<< FAILURE!
> java.lang.AssertionError: Compressed bytes contrary to configuration
> ```
> 
> Not sure why the CI was all green.

This also failed at my side. But it (DefaultCodec) is not touched by this PR. And CI passed it. Maybe it is due to environment difference?

Reply:
Removed the re-assignment.

Reply:
I think it's better to avoid this allocation if not necessary. In this case `trailerOff` is always 0 so we can just reuse the array.

Reply:
Yea, I removed the re-assignment and allocation now.